fast64_internal/f3d/f3d_node_gen.py

```diff
@@ -0,0 +1,972 @@
+    # unused in shader nodes
+    "hide_in_modifier": False,
+    "force_non_field": False,
+    "layer_selection_field": False,
+}
+
+
+class DefaultDefinition:
+    def __init__(self, names: list[str], defaults: dict[str, Any]):
+        self.names = names
+        self.defaults = GLOBAL_DEFAULTS | defaults
+
+
+DEFAULTS = [
+    DefaultDefinition(["NodeSocketFloat", "NodeSocketFloatFactor"], {"default_value": 0.0}),
+    DefaultDefinition(["NodeSocketInt"], {"default_value": 0}),
+    DefaultDefinition(["NodeSocketVector", "NodeSocketRotation"], {"default_value": (0.0, 0.0, 0.0)}),
+    DefaultDefinition(["NodeSocketColor"], {"default_value": (0.0, 0.0, 0.0, 0.0)}),
+    DefaultDefinition(["ShaderNodeMixRGB"], {"data_type": "RGBA"}),
+    DefaultDefinition(["NodeSocketVectorDirection"], {"subtype": "DIRECTION", "default_value": (0.0, 0.0, 0.0)}),
+]
+DEFAULTS = {name: definition.defaults for definition in DEFAULTS for name in definition.names}
+
+SCENE_PROPERTIES_VERSION = 2
+
+
+class ErrorState:
+    def __init__(self, error_message_queue: list[str] = None):
+        self.error_message_queue = error_message_queue or []
+        self.errors = []
+
+    def copy(self, message: str = None):
+        errors = ErrorState(self.error_message_queue.copy() + ([message] if message else []))
+        errors.errors = self.errors
+        return errors
+
+
+def print_with_exc(error_state: ErrorState, exc: Exception):
+    message = "\n".join(error_state.error_message_queue)
+    print(message + ":\n" + str(exc))
+    print(traceback.format_exc())
+    error_state.errors.append((message, exc))
+
+
+def createOrUpdateSceneProperties():
+    group = bpy.data.node_groups.get("SceneProperties")
+    upgrade_group = bool(group and group.get("version", -1) < SCENE_PROPERTIES_VERSION)
+
+    if group and not upgrade_group:
+        # Group is ready and up to date
+        return
+
+    if upgrade_group and group:
+        # Need to upgrade; remove old outputs
+        if bpy.app.version >= (4, 0, 0):
+            for item in group.interface.items_tree:
+                if item.item_type == "SOCKET" and item.in_out == "OUTPUT":
+                    group.interface.remove(item)
+        else:
+            for out in group.outputs:
+                group.outputs.remove(out)
+        new_group = group
+    else:
+        print("Creating Scene Properties")
+        # create a group
+        new_group = bpy.data.node_groups.new("SceneProperties", "ShaderNodeTree")
+        # create group outputs
+        new_group.nodes.new("NodeGroupOutput")
+
+    new_group["version"] = SCENE_PROPERTIES_VERSION
+
+    # Create outputs
+    if bpy.app.version >= (4, 0, 0):
+        tree_interface = new_group.interface
+
+        _nodeFogEnable: NodeSocketFloat = tree_interface.new_socket(
+            "FogEnable", socket_type="NodeSocketFloat", in_out="OUTPUT"
+        )
```

Comment:
could this (both versions) be refactored like
```py
    # Create outputs
    if bpy.app.version >= (4, 0, 0):
        tree_interface = new_group.interface

        def new_output_socket(name, socket_type):
            tree_interface.new_socket(name, socket_type=socket_type, in_out="OUTPUT")

    else:

        def new_output_socket(name, socket_type):
            new_group.outputs.new(socket_type, name)

    for name, socket_type in (
        ("FogEnable", "NodeSocketInt"),
        ("FogColor", "NodeSocketColor"),
        ...,
    ):
        new_output_socket(name, socket_type)
```

Reply:
If you give me a second yeah sure